### ISSUE TITLE
img HTML tag not rendered correctly most of the times.

### ISSUE THREAD

**Issue by kacperkaminski-st** [highcharts/node-export-server#413]
#### Expected behaviour

Exporting a chart where we use `img` html tag correctly fetches images and embeds them in the chart:
![image](https://github.com/highcharts/node-export-server/assets/118282321/e20b4db2-d88c-4f3f-b1c0-091d106f6a8b)

#### Actual behaviour
Most of the time some or all of the images are not rendered:
![image](https://github.com/highcharts/node-export-server/assets/118282321/16f556cf-a897-4b6f-aa46-00e173d01e27)
or:
![image](https://github.com/highcharts/node-export-server/assets/118282321/48dbfbbf-bbd2-416d-abda-179c8988b6b8)

#### Reproduction steps

`node-export-server` version:
```
❯ npm list | grep highcharts
├── highcharts-export-server@3.0.0-beta.1
```
Command used and output:
```
❯ node node_modules/highcharts-export-server/bin/cli.js --infile chart.json --constr chart --allowCodeExecution true --outfile ./chart.png --callback chart_generator/callback.js
    __  ___       __         __               __
   / / / (_)___ _/ /_  _____/ /_  ____ ______/ /______
  / /_/ / / __ `/ __ \/ ___/ __ \/ __ `/ ___/ __/ ___/
 / __  / / /_/ / / / / /__/ / / / /_/ / /  / /_(__  )
/_/ /_/_/\__, /_/ /_/\___/_/ /_/\__,_/_/   \__/____/
        ____//___/                  __     _____
       / ____/  ______  ____  _____/ /_   / ___/___  ______   _____  _____
      / __/ | |/_/ __ \/ __ \/ ___/ __/   \__ \/ _ \/ ___/ | / / _ \/ ___/
     / /____>  </ /_/ / /_/ / /  / /_    ___/ /  __/ /   | |/ /  __/ /
    /_____/_/|_/ .___/\____/_/   \__/   /____/\___/_/    |___/\___/_/
              /_/
                                                                     vundefined
Tue Aug 29 2023 21:00:28 GMT+0200 [notice] - [cache] Dependency cache is up to date, proceeding.
Tue Aug 29 2023 21:00:28 GMT+0200 [verbose] - [cache] writing new manifest
Tue Aug 29 2023 21:00:28 GMT+0200 [notice] - [browser] attempting to get a browser instance (try 0)
Tue Aug 29 2023 21:00:28 GMT+0200 [notice] - [pool] Initializing pool: min 1, max 1.
Tue Aug 29 2023 21:00:28 GMT+0200 [verbose] - [pool] Attaching exit listeners to the process.
Tue Aug 29 2023 21:00:28 GMT+0200 [notice] - [pool] Successfully created a worker 06fbf297-36cb-4ac3-be44-e0653566be6c - took 204 ms.
Tue Aug 29 2023 21:00:29 GMT+0200 [notice] - [pool] The pool is ready with 1 initial resources waiting.
Tue Aug 29 2023 21:00:29 GMT+0200 [verbose] - [chart] Starting exporting process.
Tue Aug 29 2023 21:00:29 GMT+0200 [verbose] - [chart] Attempting to export from an input file.
Tue Aug 29 2023 21:00:29 GMT+0200 [verbose] - [pool] Work received, starting to process.
Tue Aug 29 2023 21:00:29 GMT+0200 [verbose] - [pool] Aquiring worker
Tue Aug 29 2023 21:00:29 GMT+0200 [verbose] - [pool] Acquired worker handle
Tue Aug 29 2023 21:00:29 GMT+0200 [verbose] - [pool] Starting work on pool entry 06fbf297-36cb-4ac3-be44-e0653566be6c.
Tue Aug 29 2023 21:00:29 GMT+0200 [verbose] - [export] Determining export path.
Tue Aug 29 2023 21:00:29 GMT+0200 [verbose] - [export] Treating as config.
Tue Aug 29 2023 21:00:29 GMT+0200 [verbose] - [pool] Work completed in 343 ms.
Tue Aug 29 2023 21:00:29 GMT+0200 [notice] - [pool] Killing all workers.
Tue Aug 29 2023 21:00:29 GMT+0200 [verbose] - [pool] Started draining process.
Tue Aug 29 2023 21:00:29 GMT+0200 [notice] - [pool] Destroying pool entry 06fbf297-36cb-4ac3-be44-e0653566be6c.
Tue Aug 29 2023 21:00:29 GMT+0200 [verbose] - The unhandledRejection error, message: Protocol error: Connection closed. Most likely the page has been closed..
Tue Aug 29 2023 21:00:30 GMT+0200 [notice] - [pool] Killing all workers.
Tue Aug 29 2023 21:00:30 GMT+0200 [verbose] - The unhandledRejection error, message: Cannot read properties of undefined (reading 'close').
```

Content of `chart-json`:
```
{
  "title": {
    "text": "Some chart title",
    "fontSize": "64px"
  },
  "exporting": {
    "allowHTML": true
  },
  "chart": {
    "type": "column",
    "width": 800,
    "height": 600,
    "spacingBottom": 40,
    "spacingRight": 40
  },
  "xAxis": {
    "categories": [
      "http://is1-ssl.mzstatic.com/image/thumb/Purple116/v4/36/b9/cc/36b9cc99-46b4-d42c-2e15-b6bddb9cc93d/AppIcon-0-0-1x_U007emarketing-0-0-0-7-0-0-sRGB-0-0-0-GLES2_U002c0-512MB-85-220-0-0.png/30x30bb.jpg",
      "http://is1-ssl.mzstatic.com/image/thumb/Purple126/v4/b9/30/23/b93023c2-7877-e141-a0e4-766d1c0992e1/AppIcon-0-1x_U007emarketing-0-7-0-0-85-220.png/30x30bb.jpg"
    ],
    "gridLineWidth": 1
  },
  "credits": {
    "enabled": false
  },
  "yAxis": {
    "title": {
      "text": "Some data",
      "fontSize": "64px"
    }
  },
  "colorAxis": {
    "maxColor": "#0e3615",
    "minColor": "#5aec6e",
    "visible": false
  },
  "series": [
    {
      "name": "Some labels",
      "fontSize": "64px",
      "showInLegend": false,
      "data": [
        10,
        20
      ]
    }
  ]
}

```

Content of `callback/js`:
```
(chart) => {
  chart.xAxis[0].update({
    labels: {
      useHTML: true,
      padding:10,
      height:40,
      width:40,
      formatter: function () {
        const imgUrl = this.value;
        return '<img src="' + imgUrl + '"  style="width:40px; height:40px">';
      }
    }
  });
}

```

**Comment by jakubSzuminski** [comment]
Thanks for reporting @kacperkaminski-st!

You're absolutely right. I've checked your config and everything seems to be correct (this works well in Highcharts: https://jsfiddle.net/BlackLabel/bwnzrq5u/) and it should work in Export Server as well.

I'm marking it as a bug. Currently, we're working on a few prioritized issues related to server health, but after they're fixed I'll make sure to prioritize this.

As a workaround I'd recommend you try out the following approach:
1. Convert your images to `base64`
2. Instead of linking external resources, use `<img src="data:image/png;base64, iVBO...kJggg==" alt="Your alt" />` in the `labels.format` (I think you could get rid of the callback and the `formatter`).
3. Make sure to add `useHTML`.

Please let me know if it worked or if you need any help.

**Comment by kacperkaminski-st** [comment]
Hi @jakubSzuminski. Thanks for your quick answer!
Funny thing, I actually wrote whole answer with details about how I already experimented with base64 approach and it didn't work, but then tried one more thing and it works now :)

So first, some details on why base64 approach does not work out of the box. To be honest this also seems like a bug to me, but I may be missing some obvious context.

When trying it in jsfiddle the images are rendered as empty boxes: https://jsfiddle.net/4gzn8q9f/ with a warning:
```
"Highcharts warning #33: www.highcharts.com/errors/33/?Invalid%20attribute%20in%20config=src
 - Invalid attribute in config: src"
```

This seemed weird since we've already seen that it works but as as per [Highchart documentation](https://api.highcharts.com/class-reference/Highcharts.AST) I've checked that all tags and attributes that I use are supported:
```
Highcharts.AST.allowedTags.includes("img");
Highcharts.AST.allowedAttributes.includes("src");
Highcharts.AST.allowedAttributes.includes("style");
```
Result:
```
"Highcharts.AST.allowedTags.includes(\&quot;img\&quot;): ", true
"Highcharts.AST.allowedAttributes.includes(\&quot;src\&quot;): ", true
"Highcharts.AST.allowedAttributes.includes(\&quot;style\&quot;): ", true
```
Next thing to check in this case would be whether the base64 data is actually correct, but if you check this fiddle(https://jsfiddle.net/53ufo6bt/) you'll see that it's displayed correctly.

When trying to use the same code in `callback.js` and when trying to export image using export-server the image is not exported correctly as well and instead there is a red error message present.
Current `callback.js` code:
```
(chart) => {
  chart.xAxis[0].update({
    labels: {
      useHTML: true,
      padding:10,
      height:40,
      width:40,
      formatter: function () {
        return '<img src="data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAASABIAAD/4QCMRXhpZgAATU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAABJKGAAcAAAAzAAAAUKABAAMAAAABAAEAAKACAAQAAAABAAAAHqADAAQAAAABAAAAHgAAAABBU0NJSQAAADEuNzIuMS0yM0otVVZJTExIVFZIU1dFNENZNlZQNTJQUTZDT0UuMC4yLTMA/+0AOFBob3Rvc2hvcCAzLjAAOEJJTQQEAAAAAAAAOEJJTQQlAAAAAAAQ1B2M2Y8AsgTpgAmY7PhCfv/AABEIAB4AHgMBIgACEQEDEQH/xAAfAAABBQEBAQEBAQAAAAAAAAAAAQIDBAUGBwgJCgv/xAC1EAACAQMDAgQDBQUEBAAAAX0BAgMABBEFEiExQQYTUWEHInEUMoGRoQgjQrHBFVLR8CQzYnKCCQoWFxgZGiUmJygpKjQ1Njc4OTpDREVGR0hJSlNUVVZXWFlaY2RlZmdoaWpzdHV2d3h5eoOEhYaHiImKkpOUlZaXmJmaoqOkpaanqKmqsrO0tba3uLm6wsPExcbHyMnK0tPU1dbX2Nna4eLj5OXm5+jp6vHy8/T19vf4+fr/xAAfAQADAQEBAQEBAQEBAAAAAAAAAQIDBAUGBwgJCgv/xAC1EQACAQIEBAMEBwUEBAABAncAAQIDEQQFITEGEkFRB2FxEyIygQgUQpGhscEJIzNS8BVictEKFiQ04SXxFxgZGiYnKCkqNTY3ODk6Q0RFRkdISUpTVFVWV1hZWmNkZWZnaGlqc3R1dnd4eXqCg4SFhoeIiYqSk5SVlpeYmZqio6Slpqeoqaqys7S1tre4ubrCw8TFxsfIycrS09TV1tfY2dri4+Tl5ufo6ery8/T19vf4+fr/2wBDAAICAgICAgMCAgMEAwMDBAUEBAQEBQcFBQUFBQcIBwcHBwcHCAgICAgICAgKCgoKCgoLCwsLCw0NDQ0NDQ0NDQ3/2wBDAQICAgMDAwYDAwYNCQcJDQ0NDQ0NDQ0NDQ0NDQ0NDQ0NDQ0NDQ0NDQ0NDQ0NDQ0NDQ0NDQ0NDQ0NDQ0NDQ0NDQ3/3QAEAAL/2gAMAwEAAhEDEQA/ALW41p6No+q+INRh0rRrWW7uZ3WNUiQuQXIUE4HC5PJPArC31+inwDkn1L4GTaR8JNS07TPG63Ekl812ivK37w7cgqx2mLaEbayg5GAckfzLkWULH4h0ZStZN6JNu3SKbSu/U/zG4C4Qp8QZlLB1ajiowlO0UnOfLb3IJuKcnfS72TZ8V+P/AAB4h+Guur4d8TCAXjW6XIFvL5q7JCwGTgc5U8Yridxr6Y+OPjT4urotv4G+LmiWQvRP9otdYECebJEnBSKSM+XgkjftAOMBhXy7vrDN8LRoYqVPD8yj2krSXk/Tv1OHjHKsDgc1qYfL+dU19mpFxnBv7Ml1a7rR9D//0Km8+v619s+Cvg18KviJ4aj8Y/D/AMWan4ZudAtozrL3cRYwyqhd5RIrRBchWOUdlAA4B6/D2TV621TU7O3ms7S8uIILkYmiilZElGMYdVIDDB7g1/M+WYqjh5t4ikqkX01TT6NNO6137o/zF4WznAZdXlLMMJGvTktm3GSa1i4yi0462v3WnmfW37SXxY8J+K/D3hvwJ4X1KXxC2ibZLvWZ1KmeRYvK4LAFmfJZyBtzjBPOPkTefX9arZNGTUZnj6uOxDxFW13bbayVlu7/AHmHFHEWJz7MZ5ji0lJpKyvZKKSS1bb0W7bbP//Z"  style="width:40px; height:40px">';}
    }
  });
}
```
Rendered image:
![image](https://github.com/highcharts/node-export-server/assets/118282321/ddf60bdf-7026-43c1-a354-02aa5ff7a8fb)

So what solved this problem to me? Throwing security out of the window and using:
```
Highcharts.AST.bypassHTMLFiltering = true;
```
(Got this idea from [AST documentation](https://api.highcharts.com/class-reference/Highcharts.AST#.bypassHTMLFiltering)).
Please check this jsfiddle(https://jsfiddle.net/4gzn8q9f/1/) to see it working correctly.

**Comment by jakubSzuminski** [comment]
You're absolutely right, to me it seems to be a Highcharts bug (related to the `AST` module).

Could you please try out the following workaround: https://jsfiddle.net/BlackLabel/po2hdysw/ (with using the `Highcharts.SVGRenderer`)?

**API:** https://api.highcharts.com/class-reference/Highcharts.SVGRenderer

**Comment by jakubSzuminski** [comment]
@kacperkaminski-st this week we've released the Export Server v3.0.3 with many big and small fixes.

Could you please try it out and let me know if this issue still persists? If it does, I will add it to the top of our backlog and make sure we take care of it in one of the upcoming releases.

Thanks in advance!

**Comment by jakubSzuminski** [comment]
I've tried it out with the v3.0.3 version and after sending 50 requests with exactly the same config as you mentioned, 50/50 times both images have been rendered.

I'm closing this issue for now as it has probably been fixed via one of the fixes from 3.0.0 to 3.0.3.

I will reopen this in case it still does not work for you, just let me know.
